qutip/solver/sode/_noise.py

```diff
@@ -10,31 +10,84 @@ class Wiener:
     def __init__(self, t0, dt, generator, shape):
         self.t0 = t0
         self.dt = dt
-        self.generator = generator
-        self.t_end = t0
         self.shape = shape
-        self.process = np.zeros((1,) + shape, dtype=float)
+        self.generator = generator
+        self.noise = np.zeros((0,) + shape, dtype=float)
+        self.last_W = np.zeros(shape[-1], dtype=float)
+        self.idx_last_0 = 0
 
-    def _extend(self, t):
-        N_new_vals = int((t - self.t_end + self.dt*0.01) // self.dt)
+    def _extend(self, idx):
+        N_new_vals = idx - self.noise.shape[0]
         dW = self.generator.normal(
             0, np.sqrt(self.dt), size=(N_new_vals,) + self.shape
         )
-        W = self.process[-1, :, :] + np.cumsum(dW, axis=0)
-        self.process = np.concatenate((self.process, W), axis=0)
-        self.t_end = self.t0 + (self.process.shape[0] - 1) * self.dt
+        self.noise = np.concatenate((self.noise, dW), axis=0)
 
     def dW(self, t, N):
-        if t + N * self.dt > self.t_end:
-            self._extend(t + N * self.dt)
-        idx0 = int((t - self.t0 + self.dt * 0.01) // self.dt)
-        return np.diff(self.process[idx0:idx0 + N + 1, :, :], axis=0)
+        # Find the index of t.
+        # Rounded to the closest step, but only multiple of dt are expected.
+        idx0 = int((t - self.t0 + self.dt * 0.4999) // self.dt)
```

Comment:
Is there a reason not to use `round()`? Is this faster?

Reply:
I always used numpy functions and never realized there was a built in...